graphistry/plugins/spannergraph.py

```diff
@@ -0,0 +1,294 @@
+            "record_count": self.record_count
+        }
+
+
+class SpannerGraph:
+    """
+    A comprehensive interface for interacting with Google Spanner Graph databases.
+
+    :ivar str project_id: The Google Cloud project ID.
+    :ivar str instance_id: The Spanner instance ID.
+    :ivar str database_id: The Spanner database ID.
+    :ivar Any connection: The active connection to the Spanner database.
+    :ivar Any graphistry: The Graphistry parent object.
+    """
+
+    def __init__(self, graphistry: Any, project_id: str, instance_id: str, database_id: str):
+        """
+        Initializes the SpannerGraph instance.
+
+        :param graphistry: The Graphistry parent object.
+        :param project_id: The Google Cloud project ID.
+        :param instance_id: The Spanner instance ID.
+        :param database_id: The Spanner database ID.
+        """
+        self.graphistry = graphistry
+        self.project_id = project_id
+        self.instance_id = instance_id
+        self.database_id = database_id
+        self.connection = self.__connect()
+
+    def __connect(self) -> Any:
+        """
+        Establishes a connection to the Spanner database.
+
+        :return: A connection object to the Spanner database.
+        :rtype: google.cloud.spanner_dbapi.connection
+        :raises SpannerConnectionError: If the connection to Spanner fails.
+        """
+        from google.cloud.spanner_dbapi.connection import connect
+
+        try:
+            connection = connect(self.instance_id, self.database_id)
+            connection.autocommit = True
+            logging.info("Connected to Spanner database.")
+            return connection
+        except Exception as e:
+            raise SpannerConnectionError(f"Failed to connect to Spanner: {e}")
+
+    def close_connection(self) -> None:
+        """
+        Closes the connection to the Spanner database.
+        """
+        if self.connection:
+            self.connection.close()
+            logging.info("Connection to Spanner database closed.")
+
+    def execute_query(self, query: str) -> SpannerQueryResult:
+        """
+        Executes a GQL query on the Spanner database.
+
+        :param query: The GQL query to execute.
+        :return: The results of the query execution.
+        :rtype: SpannerQueryResult
+        :raises RuntimeError: If the query execution fails.
+        """
+        logger.debug(f' SpannerGraph execute_query() query:{query}\n')
+
+        try:
+            start_time = time.time()
+            cursor = self.connection.cursor()
+            cursor.execute(query)
+            results = cursor.fetchall()
+            execution_time = time.time() - start_time
+            logging.info(f"Query executed in {execution_time:.4f} seconds.")
+            return SpannerQueryResult(results, execution_time)
+        except Exception as e:
+            raise RuntimeError(f"Query execution failed: {e}")
+
+    @staticmethod
+    def convert_spanner_json(data):
```

Comment:
this to_dataframe gets tricky to maintain, having some unit tests here would help

i get the time crunch.. maybe we get at least a sample dump in comments, and follow-up pr with actual tests?